### ISSUE TITLE
Allow inclusive and exclusive ranges in alerting thresholds

### ISSUE THREAD

**Issue by gelicia** [grafana/grafana#82690]
### Discussed in https://github.com/grafana/grafana/discussions/68679

<div type='discussions-op-text'>

<sup>Originally posted by **tporeba** May 18, 2023</sup>
An issue https://github.com/grafana/grafana/issues/19193 has been closed because of moving Grafana feature requests to GitHub Discussions.

> Should this feature still be of interest to you, please re-create it in our Discussions area(https://github.com/grafana/grafana/discussions). 

So here it is, let me quote the original author:

[atlasW](https://github.com/atlasW) wrote:
> I expect the monitored value to trigger an alarm at 0. So I set the alarm condition of "is within range 0 to 0", but he did not trigger the alarm. So I looked at the code and found that'within range'does not include both sides of the boundary. So if I think it's equal to zero, call the police. How should I set it?

I completely agree. 

Currently there are following alert condition evaluators, based on [evaluator.go](https://github.com/grafana/grafana/blob/main/pkg/services/alerting/conditions/evaluator.go):

- IS ABOVE / `gt` -> strict greater than, `x > T`
- IS BELOW / `lt` -> strict lower than, `x < T`
- IS WITHIN RANGE / `within_range` -> strictly between upper and lower thresholds, `x ∈ (T1, T2)`
- IS OUTSIDE RANGE / `outside_range` -> strictly outside upper and lower thresholds, `x ∉ <T1,T2>`
- NO DATA

Why there is at least no "IS EQUAL" condition to complement those?
Is there any other non-obvious way to get equality? 
Is there a way to negate a condition? This would be awkward, but still would allow to construct equality.

I think all those non-strict versions could be useful:
- `x ≥ T`
- `x ≤ T`
- `x ∉ <T1,T2) `
- `x ∉ (T1, T2>`
- `x ∉ (T1, T2)`
- `x ∈ <T1, T2)`
- `x ∈ <T1, T2>`
- `x ∈ (T1, T2>`</div>

**Comment by teynar** [comment]
I agree as well, in fact I just found this issue after a quick Google search since I am starting to use Grafana. I even thought that the equal operator was omitted because it could be implemented as a within range condition with the same value, but I saw that it was not the case and I was left with confusion due to the fact that it looks misleading to the end user to me.

I do understand though, that [floating point numbers comparison](https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/) is a tricky subject, but integer equality is trivial: to accommodate both comparisons, maybe we could create a new operator called "is near" that would require two values: a number and an epsilon (meant for floats, but it can default to 0, and then add a tooltip somewhere in the UI that explains why). 

Also, I haven't considered changing the current implementation for "is within range" since it would break a lot of existing rules: maybe the name should be changed to something more intuitive though.

**Comment by armandgrillet** [comment]
What we're thinking about doing is having the option to set "inclusive" or "exclusive" when setting the upper and lower thresholds in  this condition:

![Image](https://github.com/grafana/grafana/assets/2117580/2076d725-4924-4b62-b8d5-2854d06faf9c)

This would allow all of these use cases:

- x ∉ <T1,T2) 
- x ∉ (T1, T2>
- x ∉ (T1, T2)
- x ∈ <T1, T2)
- x ∈ <T1, T2>
- x ∈ (T1, T2>

**Comment by yuri-tceretian** [comment]
I am not sure it makes sense to include all these use cases as drop-down items. It would be great if @gillesdemey and @dhalachliyski could weigh in on possible options. 
In the meantime, these operations can be achieved by using `Math` expression instead of `Threshold`. It could be as simple as `$A > T1 & $A <= T2`. There are some limitations, though. $A should have unique labels. Otherwise, it could cause some excessive memory consumption.

**Comment by chenlujjj** [comment]
Any updates ?
I think at least the behavior should be documented to avoid confusion
